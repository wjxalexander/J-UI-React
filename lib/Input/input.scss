@import "../_helper";

.j-ui-input {
  box-shadow: none;
  box-sizing: border-box;
  cursor: text;
  display: inline-block;
  flex-direction: row;
  align-items: stretch;
  position: relative;
  border: 1px solid $custom-border-color;
  border-image: initial;
  border-radius: 2px;
  background: rgb(255, 255, 255);


  -webkit-font-smoothing: antialiased;
  font-size: 1em;
  font-weight: 400;
  padding: 0px 8px;
  color: rgb(50, 49, 48);
  width: 100%;
  min-width: 0px;
  text-overflow: ellipsis;
  line-height: 30px;

  &:hover{
    border-color: rgb(50, 49, 48)
  }
  &:focus{
    border-color:$custom-focus-color;
    outline: none;

  }

}
