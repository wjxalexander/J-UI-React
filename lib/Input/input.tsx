import React, {useRef, useState} from 'react';
import cls from "classnames";
import "./input.scss"
import Icon, {IconProps} from "../Icon/icon";

interface InputProps extends React.AllHTMLAttributes<HTMLInputElement> {
  className?: string;
  mask?: maskProps;
  componentRef?: React.RefObject<HTMLInputElement>;
  onChange?: (e: React.ChangeEvent<HTMLInputElement>, newValue?: string) => void;
  iconProps?:IconProps
}

interface maskProps {
  regExp: RegExp;
}

function setSelectionRange(input: HTMLInputElement, selectionStart: number, selectionEnd: number) {
  input.focus();
  input.setSelectionRange(selectionStart, selectionEnd)
}


const Input: React.FunctionComponent<InputProps> = (props: InputProps) => {
  const {className, mask, componentRef, onChange,iconProps, ...restProps} = props;
  const textInput = useRef<HTMLInputElement>(null);
  const value = props.value || "";
  const [inputVal, setInputVal] = useState(value);
  const _onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newVal = e.target.value;
    if (!mask) {
      setInputVal(newVal);
      props.onChange && props.onChange(e)
    } else if (mask && typeof inputVal === "string") {
      const globalsRegExp = mask.regExp;
      const originalLength: number = inputVal.length;
      new Promise((resolve, reject) => {
        if (newVal.length > inputVal.length) {
          for (let i = 0; i < originalLength; i++) {
            if (newVal[i] !== inputVal[i]) {
              if (/[\d_]/.test(newVal[i])) {
                let newVal2 = newVal.split("");
                newVal2.splice(i + 1, 1);
                const newVal3 = newVal2.join("");
                if (globalsRegExp.test(newVal3)) {
                  setInputVal(newVal3);
                  onChange && onChange(e, newVal3);
                }
                resolve(i);
                break;
              } else {
                reject(i);
              }
            }
          }
        } else if (newVal.length < originalLength) {
          if (!textInput.current || !textInput.current.selectionEnd) {
            return
          } else {
            const position = textInput.current.selectionEnd;
            let arraySplit = [...(inputVal as string).split("")];
            if (/[\d_]/.test((inputVal as string)[position])) {
              arraySplit.splice(position, 1, "_");
              setInputVal(arraySplit.join(""));
              onChange && onChange(e, arraySplit.join(""));
              resolve(position - 1);
            } else {
              reject(position)
            }
          }
        }
      }).then((i: number) => {
        setSelectionRange(textInput.current!, i + 1, i + 1);
      })
        .catch(i => {
          setSelectionRange(textInput.current!, i, i);
        });
    }
  };
  return (
    <span className="J-UI-input-wrapper">
      <input ref={textInput} value={inputVal} onChange={_onChange}
                className={cls(className, 'J-UI-input')}{...restProps} />
      {iconProps && iconProps.iconName&&<Icon {...iconProps} />}
    </span>)

};

Input.defaultProps = {};


export default Input;