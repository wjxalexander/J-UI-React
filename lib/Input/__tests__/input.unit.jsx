import React from "react";
import Input from '../input';
import Enzyme, {mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
// import {Icon, Button} from "../../index"
import focusTest from "../../__tests__/focusTest"

Enzyme.configure({adapter: new Adapter()});
describe("render input Proper",()=>{
  // focusTest(Input);
  let container;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  afterEach(() => {
    document.body.removeChild(container);
  });
  it("mount correct",()=>{
    const input = mount(<Input />);
    expect(input).toMatchSnapshot()
  });
  it('focus() and onFocus', () => {
    const handleFocus = jest.fn();
    const wrapper = mount(<Input onFocus={handleFocus}/>, {attachTo: container});
    expect(wrapper.prop('onFocus')).toBeTruthy();
    // console.log(wrapper.debug());
    wrapper.find('input').simulate('focus');
    expect(handleFocus).toHaveBeenCalled();
  });
  it('blur() and onBlur', () => {
    const handleBlur = jest.fn();
    const wrapper = mount(<Input onBlur={handleBlur}/>, {attachTo: container});
    expect(wrapper.prop('onBlur')).toBeTruthy();
    wrapper.find('input').simulate('blur');
    expect(handleBlur).toHaveBeenCalled();
  });
  it('autoFocus', () => {
    const handleFocus = jest.fn();
    mount(<Input autoFocus onFocus={handleFocus}/>, {attachTo: container});
    // jest.runAllTimers();
    expect(handleFocus).toHaveBeenCalled();
  });
  it('should support maxLength', () => {
    const wrapper = mount(<Input maxLength={3} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should support Icons', () => {

    const wrapper = mount(<Input maxLength={3} iconProps={{iconName:"Calendar",USEMsFabricIcon:true}} />);
    expect(wrapper).toMatchSnapshot();
  });
})