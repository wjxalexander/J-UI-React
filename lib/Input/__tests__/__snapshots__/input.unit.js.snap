// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render input Proper mount correct 1`] = `
<Input>
  <span
    className="J-UI-input-wrapper"
  >
    <input
      className="J-UI-input"
      onChange={[Function]}
      value=""
    />
  </span>
</Input>
`;
